### ISSUE TITLE
Feature Request: Equirectangular (360 Video) Output for VR Playback

### ISSUE THREAD

**Issue by Cat5TV** [minetest/minetest#6099]
I would like to propose / request the addition of equirectangular video output for the Minetest client.

Equirectangular video, in a nutshell, is a full 360x360-degree view rendered in a 3840x1920 output. This output can be viewed using VR headsets such as Oculus Rift, Samsung Gear VR, or via a smartphone using [a special headset / Google Cardboard](https://cat5.tv/cardboard).

Sample of 360 Video from Equirectangular source (set to 4K if your Internet can handle it, and drag your mouse on video): [https://category5.tv/shows/technology/episode/471-360/](https://category5.tv/shows/technology/episode/471-360/)

Sample stillshot of same video (to show what Equirectangular video _actually_ looks like before being processed by the player): [https://i.vimeocdn.com/video/597592917.png](https://i.vimeocdn.com/video/597592917.png)

**Additional Request:** I would imagine it could be challenging to capture equirectangular video with standard capture software, so if Minetest could render the video to disk (mJPEG would be big files but very light on the CPU, so would be my recommendation) that'd be even better.

**Advantages:**
- Shows such as [The Pixel Shadow](https://minetest.tv) could begin shooting episodes which would playback in 360 space. This means the viewer could literally look around within the video. This would set Minetest apart and really make YouTube videos standout. I imagine we'd very quickly see Minecraft players asking "How did they do this?"
- An unmatched way for mod developers to showcase their mods within VR space.
- This could lead to more creativity within the community. I can imagine folks putting together gameplay "movies" that push the limits of current creativity since the viewer can literally look around the video.
- Game players could utilize this with their VR headsets to become immersed more fully in Minetest.

360 Video has proven it is not a fad: it is here to stay. Where it goes next is up to content creators. I personally cannot wait to use Minetest as a platform for creativity in developing an entire online video series featuring Minetest in 360-video through equirectangular 3840x1920 output.

Robbie Ferguson, President
Category5 TV Network
[www.Category5.TV](https://Category5.TV)

**Comment by Ezhh** [comment]
Would love to see this. :)

**Comment by C1ffisme** [comment]
I want to see where this goes.

**Comment by Brackston** [comment]
Nothing like a well rounded point of view. :) 
Great Idea, this would certainly set us apart.

**Comment by paramat** [comment]
Would be intensive to render, as currently we only render the world contained within the view frustrum, you would need a fast computer to preserve FPS.
This also might be quite a complex task to make possible as MT is probably designed around a limited view frustrum.

I have tried out these videos and they are interesting gimmicks for the first couple of times, but visually distorted and the novelty wears off quickly. I feel that most video viewers want to be passive and not have to drag their view around, if they want view control they can play the game as a player.

> Game players could utilize this with their VR headsets to become immersed more fully in Minetest.

Players can already do this with our existing 3D modes. The player determines the view direction and MT renders what is in that direction instead of rendering a complete surrounding. There is no need at all for full-surround for a player, it is only for passive viewers who are not in control of the player.

So a complex task plus the code that then needs mantaining, just for a gimmick for passive viewers doesn't seem justified to me, especially since we have an extreme lack of dev time at the moment, better that devs use their limited time for more useful and higher priority things. Sorry for my opinion.

**Comment by Cat5TV** [comment]
> Would be intensive to render, as currently we only render the world contained within the view frustrum, you would need a fast computer to preserve FPS.

Of course. This is to be expected. Can Minetest render via GPU? If yes, that would help. Keep in mind the mJPEG output as described would be very light on CPU, so the only overhead would be that of the rendering engine itself. That said, we wouldn't mind throwing our most powerful systems at this.

>  I feel that most video viewers want to be passive and not have to drag their view around

360 viewers tend to use headsets. The "mouse drag" is possible for those who do not have one, but is not the intended audience (more of a backward compatibility). In VR you simply move your head.

> players can already do this with our existing 3D modes

Yes, but not in recordings. To be clear, Equirectangular output would be for recordings as described in the first post. Whether or not it is used for gameplay doesn't really matter to me.

> just for a gimmick for passive viewers (non-players) doesn't seem justified to me

I understand your point re. limited dev time, and I appreciate that fully. However, I do want to reiterate that 360 video has proven itself not to be "just a gimmick" - there are entire companies starting up to enhance the platform:

- YouTube and Vimeo both support _and promote_ Equirectangular.
- VeeR VR is 100% dedicated to equirectangular (they do not accept anything but 360 video).
- Facebook supports equirectangular and even bought Oculus so they can build the VR experience.
- A viewer can invest as little as $20 now for a good quality Google Cardboard headset and experience true VR.

It is also worth mentioning that Minecraft got HUGE thanks to YouTubers. (And vice-versa, incidentally). But "gimmicks" are precisely what's needed to make Minetest stand-out in that space. Something like VR view for playback would be an exceptional boost to YouTubers who want to feature Minetest gameplay, and due to its open source nature, I'm confident there would be developer support from those truly interested in pushing the envelope. I'd be right there if I knew how to code this stuff, but as it is I offer myself as an experienced broadcaster who has a good amount of experience working with 360 video.

Our network is ready to stand behind 360 video and promote Minetest through the medium.

**Comment by rubenwardy** [comment]
I personally own a VR headset, and have watched 360 videos

**Comment by raymoo** [comment]
I don't think it would be too hard to go from one frustum to an equirectangular thing. Just render for the 6 faces of a cube map and then map each pixel of the equirectangular output to wherever in the cubemap it needs to look.

**Comment by Cat5TV** [comment]
**Update:** Windows 10's "Movies & TV" app has been updated to support playback of Equirectangular 360 Video. All Windows 10 users can now playback these videos with the OS' built-in media player.

**Comment by srifqi** [comment]
Well, why not? #8475

**Comment by farooqkz** [comment]
This issue is from another decade, and some considered that computers are not powerful enough. But now in the next decade, there are even quite unexpensive computers with iGPUs which would be able to render such a thing, possibly without special effects.
